File: ios/GradeThread/Marketplaces/Negotiation/NegotiationService.swift
import Foundation

/// Edge client for eBay best-offers + buyer messages (US-673). All calls hit the
/// consolidated edge service, which resolves the caller's eBay token, so there's
/// no cross-tenant surface. Behind a protocol so ``NegotiationStore`` is testable.
protocol NegotiationProviding {
    func offers() async throws -> [BestOffer]
    func respond(bestOfferId: String, itemId: String, action: String, counterPrice: Double?, message: String?) async throws
    /// US-1967: cheap capability probe — can this connection send offers at all?
    func capability() async throws -> NegotiationCapability
    func eligibleItems() async throws -> [EligibleNegotiationItem]
    func sendOffer(listingIds: [String], discountPercentage: String, message: String?) async throws
    func messages() async throws -> [BuyerMessage]
    func reply(messageId: String, itemId: String, recipientId: String, body: String) async throws
}

struct NegotiationService: NegotiationProviding {
    private struct OffersResponse: Decodable { let offers: [BestOffer] }
    private struct MessagesResponse: Decodable { let messages: [BuyerMessage] }
    private struct EligibleResponse: Decodable { let items: [EligibleNegotiationItem] }
    private struct OKResponse: Decodable { let ok: Bool? }

    func offers() async throws -> [BestOffer] {
        let resp: OffersResponse = try await EdgeAPI.shared.getJSON("/api/flipdesk/ebay/negotiation/offers")
        return resp.offers
    }

    func respond(
        bestOfferId: String, itemId: String, action: String,
        counterPrice: Double?, message: String?
    ) async throws {
        struct Body: Encodable {
            let itemId: String
            let action: String
            let counterPrice: Double?
            let message: String?
        }
        let _: OKResponse = try await EdgeAPI.shared.postJSON(
            "/api/flipdesk/ebay/negotiation/offers/\(bestOfferId)/respond",
            body: Body(itemId: itemId, action: action, counterPrice: counterPrice, message: message)
        )
    }

    func capability() async throws -> NegotiationCapability {
        try await EdgeAPI.shared.getJSON("/api/flipdesk/ebay/negotiation/capabilities")
    }

    func eligibleItems() async throws -> [EligibleNegotiationItem] {
        let resp: EligibleResponse = try await EdgeAPI.shared.getJSON("/api/flipdesk/ebay/negotiation/eligible")
        return resp.items
    }

    func sendOffer(listingIds: [String], discountPercentage: String, message: String?) async throws {
        struct Body: Encodable {
            let listingIds: [String]
            let discountPercentage: String
            let message: String?
        }
        let _: OKResponse = try await EdgeAPI.shared.postJSON(
            "/api/flipdesk/ebay/negotiation/send-offer",
            body: Body(listingIds: listingIds, discountPercentage: discountPercentage, message: message)
        )
    }

    func messages() async throws -> [BuyerMessage] {
        let resp: MessagesResponse = try await EdgeAPI.shared.getJSON("/api/flipdesk/ebay/messages")
        return resp.messages
    }

    func reply(messageId: String, itemId: String, recipientId: String, body: String) async throws {
        struct Body: Encodable {
            let itemId: String
            let recipientId: String
            let body: String
        }
        let _: OKResponse = try await EdgeAPI.shared.postJSON(
            "/api/flipdesk/ebay/messages/\(messageId)/reply",
            body: Body(itemId: itemId, recipientId: recipientId, body: body)
        )
    }
}

File: android/app/src/main/java/com/gradethread/app/marketplaces/negotiation/NegotiationService.kt
package com.gradethread.app.marketplaces.negotiation

import com.gradethread.app.platform.net.EdgeApi
import com.gradethread.app.platform.net.EdgeApiError
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton

/**
 * US-1354: the edge client for eBay best offers, seller-initiated offers and
 * buyer messages.
 *
 * Every call resolves the caller's eBay token server-side, so there is no
 * cross-tenant surface here — the client never names an account.
 */
@Singleton
class NegotiationService @Inject constructor(
    @Named("shared") private val edge: EdgeApi,
) {

    companion object {
        private const val BASE = "/api/flipdesk/ebay"
        const val OFFERS_PATH = "$BASE/negotiation/offers"
        const val CAPABILITIES_PATH = "$BASE/negotiation/capabilities"
        const val ELIGIBLE_PATH = "$BASE/negotiation/eligible"
        const val SEND_OFFER_PATH = "$BASE/negotiation/send-offer"
        const val MESSAGES_PATH = "$BASE/messages"

        fun respondPath(bestOfferId: String) = "$OFFERS_PATH/$bestOfferId/respond"
        fun replyPath(messageId: String) = "$MESSAGES_PATH/$messageId/reply"
    }

    @Serializable
    private data class RespondRequest(
        @SerialName("item_id") val itemId: String,
        val action: String,
        @SerialName("counter_price") val counterPrice: Double? = null,
        val message: String? = null,
    )

    @Serializable
    private data class SendOfferRequest(
        @SerialName("listing_ids") val listingIds: List<String>,
        @SerialName("discount_percentage") val discountPercentage: String,
        val message: String? = null,
    )

    @Serializable
    private data class ReplyRequest(
        @SerialName("item_id") val itemId: String,
        @SerialName("recipient_id") val recipientId: String,
        val body: String,
    )

    suspend fun offers(): List<BestOffer> =
        edge.json.decodeFromString(OffersResponse.serializer(), edge.getRaw(OFFERS_PATH)).offers

    suspend fun messages(): List<BuyerMessage> =
        edge.json
            .decodeFromString(MessagesResponse.serializer(), edge.getRaw(MESSAGES_PATH))
            .messages

    /**
     * Cheap by design — the endpoint reads a scope list and a flag and never
     * calls eBay, so the inbox can probe it on every open.
     */
    suspend fun capability(): NegotiationCapability =
        edge.json.decodeFromString(
            NegotiationCapability.serializer(),
            edge.getRaw(CAPABILITIES_PATH),
        )

    suspend fun eligibleItems(): List<EligibleItem> =
        edge.json.decodeFromString(EligibleResponse.serializer(), edge.getRaw(ELIGIBLE_PATH)).items

    /**
     * Accept, counter or decline. [counterPrice] is required for a counter and
     * ignored otherwise — the server rejects a non-positive one.
     */
    suspend fun respond(
        bestOfferId: String,
        itemId: String,
        action: OfferAction,
        counterPrice: Double? = null,
        message: String? = null,
    ) {
        edge.postRaw(
            respondPath(bestOfferId),
            edge.json.encodeToString(
                RespondRequest.serializer(),
                RespondRequest(
                    itemId = itemId,
                    action = action.wire,
                    counterPrice = counterPrice.takeIf { action == OfferAction.COUNTER },
                    message = message?.trim()?.takeIf { it.isNotEmpty() },
                ),
            ),
        )
    }

    suspend fun sendOffer(listingIds: List<String>, discountPercent: Int, message: String? = null) {
        edge.postRaw(
            SEND_OFFER_PATH,
            edge.json.encodeToString(
                SendOfferRequest.serializer(),
                SendOfferRequest(
                    listingIds = listingIds,
                    discountPercentage = NegotiationRules.discountWire(discountPercent),
                    message = message?.trim()?.takeIf { it.isNotEmpty() },
                ),
            ),
        )
    }

    suspend fun reply(messageId: String, itemId: String, recipientId: String, body: String) {
        edge.postRaw(
            replyPath(messageId),
            edge.json.encodeToString(
                ReplyRequest.serializer(),
                ReplyRequest(itemId = itemId, recipientId = recipientId, body = body.trim()),
            ),
        )
    }

    /** User-facing copy for a failed call. */
    fun message(error: Throwable): String =
        (error as? EdgeApiError)?.userMessage()
            ?: error.message
            ?: "Something went wrong talking to eBay."
}
